enterprise/reporting/flusher/flusher.go

```diff
@@ -0,0 +1,348 @@
+		return currentUTC.Sub(start)
+	}
+}
+
+func (f *Flusher) FlushAggressively(ctx context.Context) bool {
+	if !f.aggressiveFlushEnabled.Load() {
+		return false
+	}
+	lag := f.getLag(ctx)
+	reportingLagInMins := lag.Minutes()
+	return reportingLagInMins > f.lagThresholdForAggresiveFlushInMins.Load().Minutes()
+}
+
+// flush is the main logic for flushing data.
+func (f *Flusher) Flush(ctx context.Context) error {
+	currentUTC := time.Now().UTC()
+
+	// Emit the lag metric
+	lag := f.getLag(ctx)
+	f.flushLag.Gauge(lag.Seconds())
+
+	// 1. Get the time range to flush
+	s := time.Now()
+	start, end, valid, err := f.getRange(ctx, currentUTC)
+	if err != nil {
+		return err
+	}
+
+	if !valid {
+		return nil
+	}
+	f.minReportedAtQueryTimer.Since(s)
+
+	// 2. Aggregate reports. We have different aggregators for in-app and in-db aggregation
+	s = time.Now()
+	jsonReports, err := f.aggregate(ctx, start, end)
+	if err != nil {
+		return err
+	}
+	f.aggReportsTimer.Since(s)
+
+	// 3. Flush aggregated reports
+	s = time.Now()
+	err = f.send(ctx, jsonReports)
+	if err != nil {
+		return err
+	}
+	f.sendReportsTimer.Since(s)
+
+	// 4. Delete reports
+	s = time.Now()
+	if err := f.delete(ctx, start, end); err != nil {
+		return err
+	}
+	f.deleteReportsTimer.Since(s)
+
+	return nil
+}
+
+// Since we have hourly/daily/monthly aggregates on Reporting Service, we want the window to be within same hour
+// Don't consider most recent data where there are inserts happening
+// Always try to consider full window of flush interval or till current hour
+func (f *Flusher) getRange(ctx context.Context, currentUTC time.Time) (start, end time.Time, valid bool, err error) {
+	start, err = f.getStart(ctx)
+	if err != nil {
+		return time.Time{}, time.Time{}, false, err
+	}
+
+	if start.IsZero() {
+		return start, end, false, nil
+	}
+
+	end = start.Add(f.flushWindow.Load())
+	nextHour := start.Truncate(time.Hour).Add(time.Hour)
+	endLimit := currentUTC.Add(-f.recentExclusionWindow.Load())
+
+	if end.After(nextHour) {
+		end = nextHour
+	}
+	if end.After(endLimit) {
+		end = endLimit
+	}
+
+	if end.Sub(start) == f.flushWindow.Load() || end == nextHour {
+		return start, end, true, nil
+	}
+
+	return start, end, false, nil
+}
+
+func (f *Flusher) aggregate(ctx context.Context, start, end time.Time) ([]json.RawMessage, error) {
+	jsonReports, err := f.aggregator.Aggregate(ctx, start, end)
+	if err != nil {
+		return nil, err
+	}
+
+	return jsonReports, nil
+}
+
+func (f *Flusher) getConcurrency(ctx context.Context) int {
+	if f.FlushAggressively(ctx) {
+		return f.maxConcurrentRequests.Load()
+	}
+	return f.minConcurrentRequests.Load()
+}
+
+func (f *Flusher) send(ctx context.Context, aggReports []json.RawMessage) error {
+	batchSize := f.batchSizeToReporting.Load()
+	concurrency := f.getConcurrency(ctx)
+
+	g, ctx := errgroup.WithContext(ctx)
+	g.SetLimit(concurrency)
+	f.concurrentRequests.Gauge(float64(concurrency))
+
+	for i := 0; i < len(aggReports); i += batchSize {
```

Comment:
you can use lo.chunk https://github.com/samber/lo/blob/d93dd9a2d7d22037f0c301f8d91ac79f582477c4/slice.go#L168